x/sequencer/keeper/rotation.go

```diff
@@ -0,0 +1,122 @@
+package keeper
+
+import (
+	"sort"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dymensionxyz/dymension/v3/x/sequencer/types"
+)
+
+func (k Keeper) startNoticePeriodForSequencer(ctx sdk.Context, seq *types.Sequencer) time.Time {
+	completionTime := ctx.BlockHeader().Time.Add(k.NoticePeriod(ctx))
+	seq.UnbondTime = completionTime
+
+	k.UpdateSequencer(ctx, *seq) // only bonded sequencers can have notice period
+	k.SetNoticePeriodQueue(ctx, *seq)
+
+	nextSeq := k.ExpectedNextProposer(ctx, seq.RollappId)
+	if nextSeq.SequencerAddress == "" {
+		k.Logger(ctx).Info("rollapp will be left with no proposer after notice period", "rollappId", seq.RollappId, "sequencer", seq.SequencerAddress)
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeNoticePeriodStarted,
+			sdk.NewAttribute(types.AttributeKeySequencer, seq.SequencerAddress),
+			sdk.NewAttribute(types.AttributeKeyNextProposer, nextSeq.SequencerAddress),
+			sdk.NewAttribute(types.AttributeKeyCompletionTime, completionTime.String()),
+		),
+	)
+
+	return completionTime
+}
+
+// MatureSequencersWithNoticePeriod moves all the sequencers that have finished their notice period
+func (k Keeper) MatureSequencersWithNoticePeriod(ctx sdk.Context, currTime time.Time) {
+	seqs := k.GetMatureNoticePeriodSequencers(ctx, currTime)
+	for _, seq := range seqs {
+		k.StartRotation(ctx, seq.RollappId)
+		k.removeNoticePeriodSequencer(ctx, seq)
+	}
+}
+
+// IsRotating returns true if the rollapp is currently rotating proposers
+func (k Keeper) IsRotating(ctx sdk.Context, rollappId string) bool {
+	return k.HasNextProposer(ctx, rollappId)
+}
+
+// IsNoticePeriodRequired returns true if the sequencer requires a notice period before unbonding
+func (k Keeper) IsNoticePeriodRequired(ctx sdk.Context, seq types.Sequencer) bool {
+	return k.IsProposer(ctx, seq.RollappId, seq.SequencerAddress) || k.IsNextProposer(ctx, seq.RollappId, seq.SequencerAddress)
+}
+
+// ExpectedNextProposer returns the next proposer for a rollapp
+func (k Keeper) ExpectedNextProposer(ctx sdk.Context, rollappId string) types.Sequencer {
+	seqs := k.GetSequencersByRollappByStatus(ctx, rollappId, types.Bonded)
+	if len(seqs) == 0 {
+		return types.Sequencer{}
+	}
+
+	// take the next bonded sequencer to be the proposer. sorted by bond
+	sort.SliceStable(seqs, func(i, j int) bool {
+		return seqs[i].Tokens.IsAllGT(seqs[j].Tokens)
+	})
+
+	// filter out proposer and nextProposer
+	active, _ := k.GetProposer(ctx, rollappId)
+	next, _ := k.GetNextProposer(ctx, rollappId)
+	for _, s := range seqs {
+		if s.SequencerAddress == active.SequencerAddress || s.SequencerAddress == next.SequencerAddress {
+			continue
+		}
+		return s
+	}
```

Comment:
it was to avoid a case where
* seqA finishes notice period,
* nextSeq = seqB
and then seqB starts it's notice period (before rotation completed)

the idea was that if nextSeq is set, it means the rotation is in progress.
calling `ExpectedNextProposer` should give the **next** proposer, after the current rotation in progress